cirq/google/op_serializer.py

```diff
@@ -200,3 +279,74 @@ def _check_type(self, value: arg_func_langs.ARG_LIKE, arg: SerializingArg) -> No
                     arg.serialized_name, arg.serialized_type, type(value)
                 )
             )
+
+
+class CircuitOpSerializer(OpSerializer):
+    """Describes how to serialize CircuitOperations."""
+
+    @property
+    def internal_type(self):
+        return circuits.FrozenCircuit
+
+    @property
+    def serialized_id(self):
+        return 'circuit'
+
+    @property
+    def can_serialize_predicate(self):
+        return lambda op: isinstance(op.untagged, circuits.CircuitOperation)
+
+    def to_proto(
+        self,
+        op: 'cirq.CircuitOperation',
+        msg: Optional[v2.program_pb2.CircuitOperation] = None,
+        *,
+        arg_function_language: Optional[str] = '',
+        constants: List[v2.program_pb2.Constant] = None,
+        raw_constants: List[Any] = None,
+    ) -> Optional[v2.program_pb2.CircuitOperation]:
+        """Returns the cirq.google.api.v2.CircuitOperation message as a proto dict.
+
+        Note that this function requires the constants and raw_constants lists
+        to be pre-populated with the circuit in op.
+        """
+        if constants is None or raw_constants is None:
+            raise ValueError(
+                'CircuitOp serialization requires a constants list and a corresponding list of '
+                'pre-serialization values (raw_constants).'
+            )
+
+        if not isinstance(op, circuits.CircuitOperation):
+            raise ValueError(f'Serializer expected CircuitOperation but got {type(op)}.')
+
+        msg = msg or v2.program_pb2.CircuitOperation()
+        try:
+            msg.circuit_constant_index = raw_constants.index(op.circuit)
```

Comment:
Any idea if this will be a performance concern?  It seems we are doing a O(n) search of comparisons of potentially big circuits here.  I am not sure how fast comparing two circuits would be.  Maybe this is not a huge concern since a circuit would likely not have _too_ many circuitsops, but this does seem to be a potential performance issue.

Do you think it is worth making raw_constants a dictionary instead?

Reply:
That's a legitimate concern, although I would hope proper use of CircuitOperations would limit the size of these circuits somewhat. Since `op.circuit` is guaranteed to be a hashable `FrozenCircuit`, I'll change this to a constant-to-index map.

Reply:
Switched to a map. Since this broke the symmetry between `raw_constants` in the serializer and deserializer, I went ahead with the rename to `deserialized_constants` in the deserializer as well.